catalog/async/catalog-async-processingframework/src/main/java/org/codice/ddf/catalog/async/processingframework/impl/InMemoryProcessingFramework.java

```diff
@@ -0,0 +1,336 @@
+                                        .getCanonicalName(),
+                                e);
+                    }
+                }
+            });
+        }
+    }
+
+    private <T extends ProcessResourceItem> void closeInputStream(ProcessRequest<T> request) {
+        request.getProcessItems()
+                .stream()
+                .map(ProcessResourceItem::getProcessResource)
+                .filter(Objects::nonNull)
+                .forEach(resource -> {
+                    try {
+                        resource.getInputStream()
+                                .close();
+                    } catch (IOException e) {
+                        LOGGER.debug(
+                                "Failed to stream of process request. There may be temporary files left in data/tmp.");
+                    }
+                });
+    }
+
+    private <T extends ProcessResourceItem> void storeProcessRequest(
+            ProcessRequest<T> processRequest) {
+        LOGGER.trace("Storing update request post processing change(s)");
+
+        Map<String, ContentItem> contentItemsToUpdate = new HashMap<>();
+        Map<String, Metacard> metacardsToUpdate = new HashMap<>();
+
+        List<TemporaryFileBackedOutputStream> tfbosToCleanUp = new ArrayList<>();
+        for (T item : processRequest.getProcessItems()) {
+            if (item.getProcessResource() == null && item.isMetacardModified()) {
+                metacardsToUpdate.put(item.getMetacard()
+                        .getId(), item.getMetacard());
+            }
+
+            final ProcessResource processResource = item.getProcessResource();
+            TemporaryFileBackedOutputStream tfbos = null;
+            if (processResource != null && processResource.isModified()
+                    && !contentItemsToUpdate.containsKey(item.getMetacard()
+                    .getId() + processResource.getQualifier())) {
+                try {
+                    tfbos = new TemporaryFileBackedOutputStream();
+                    long numberOfBytes = IOUtils.copyLarge(processResource.getInputStream(), tfbos);
+                    LOGGER.debug("Copied {} bytes to TemporaryFileBackedOutputStream.",
+                            numberOfBytes);
+                    ByteSource byteSource = tfbos.asByteSource();
+
+                    ContentItem contentItem = new ContentItemImpl(item.getMetacard()
+                            .getId(),
+                            processResource.getQualifier(),
+                            byteSource,
+                            processResource.getMimeType(),
+                            processResource.getName(),
+                            processResource.getSize(),
+                            item.getMetacard());
+
+                    contentItemsToUpdate.put(contentItem.getId() + contentItem.getQualifier(),
+                            contentItem);
+                    tfbosToCleanUp.add(tfbos);
+                } catch (IOException e) {
```

Comment:
❓ Just thought of something. Maybe we should also catch `RuntimeException` here to make sure we clean things up even when something unexpected happens.